### PR TITLE
KAZOO-4365 - webhook updates

### PR DESCRIPTION
**Pull request by jamesaimonetti**

### CODE REVIEW

**Comment by jamesaimonetti** [comment]
@lazedo ok, PR is ready for review

File: applications/crossbar/src/modules/cb_webhooks.erl [hunk 1]
```diff
@@ -59,6 +61,56 @@ init() ->
...
+
+-spec maybe_revise_schema() -> 'ok'.
+-spec maybe_revise_schema(wh_json:object()) -> 'ok'.
+-spec maybe_revise_schema(wh_json:object(), ne_binary()) -> 'ok'.
+maybe_revise_schema() ->
+    case wh_json_schema:load(<<"webhooks">>) of
+        {'ok', SchemaJObj} ->
+            maybe_revise_schema(SchemaJObj);
+        {'error', _E} ->
+            lager:warning("failed to find webhooks schema: ~o", [_E])
```
Comment (fenollp) [inline applications/crossbar/src/modules/cb_webhooks.erl]: What's `~o`?

File: applications/crossbar/src/modules/cb_webhooks.erl [hunk 2]
```diff
@@ -59,6 +61,56 @@ init() ->
...
+    end.
+
+maybe_revise_schema(SchemaJObj) ->
+    case whapps_util:get_master_account_db() of
+        {'ok', MasterDb} -> maybe_revise_schema(SchemaJObj, MasterDb);
+        {'error', _E} ->
+            lager:warning("fmaster account not set yet, unable to revise schema: ~p", [_E])
```
Comment (fenollp) [inline applications/crossbar/src/modules/cb_webhooks.erl]: `fmaster`?

File: applications/crossbar/src/cb_context.erl [hunk 3]
```diff
@@ -613,7 +613,13 @@ validate_request_data(Schema, Context, OnSuccess, OnFailure) ->
 -spec failed(context(), jesse_error:error_reasons()) -> context().
 -spec failed_error(jesse_error:error_reason(), context()) -> context().
 failed(Context, Errors) ->
-    lists:foldl(fun failed_error/2, set_resp_status(Context, 'error'), Errors).
+    lists:foldl(fun failed_error/2
+               ,set_resp_error_code(
+                  set_resp_status(Context, 'error')
+                  ,400
+                 )
```
Comment (fenollp) [inline applications/crossbar/src/cb_context.erl]: Do you mind using `setters/2`, before the fold? (I'm not a fan of function calls spanning 7 lines…)

File: core/whistle_couch-1.0.0/src/couch_util.erl [hunk 4]
```diff
@@ -92,13 +92,13 @@
 
 -type ddoc() :: ne_binary() | 'all_docs' | 'design_docs'.
 
--type db_classifications() :: 'account' | 'modb' | 'acdc' |
+-type db_classification() :: 'account' | 'modb' | 'acdc' |
                               'numbers' | 'aggregate' | 'system' |
                               'deprecated' | 'undefined'.
```
Comment (fenollp) [inline core/whistle_couch-1.0.0/src/couch_util.erl]: Indentation

**Comment by fenollp** [comment]
On 18.1:
```
warn_behaviour                webhooks_sup.erl:84: The return type {'ok',{{'one_for_all',non_neg_integer(),0} | {'one_for_one',non_neg_integer(),0} | {'rest_for_one',non_neg_integer(),0} | {'simple_one_for_one',non_neg_integer(),0},[{_,{atom() | tuple(),atom(),'undefined' | [any()]},'permanent' | 'temporary' | 'transient','brutal_kill' | 'infinity' | non_neg_integer(),'supervisor' | 'worker','dynamic' | [atom() | tuple()]} | #{}]}} in the specification of init/1 is not a subtype of 'ignore' | {'ok',{{'one_for_all',non_neg_integer(),pos_integer()} | {'one_for_one',non_neg_integer(),pos_integer()} | {'rest_for_one',non_neg_integer(),pos_integer()} | {'simple_one_for_one',non_neg_integer(),pos_integer()} | #{},[{_,{atom() | tuple(),atom(),'undefined' | [any()]},'permanent' | 'temporary' | 'transient','brutal_kill' | 'infinity' | non_neg_integer(),'supervisor' | 'worker','dynamic' | [atom() | tuple()]} | #{}]}}, which is the expected return type for the callback of supervisor behaviour
```

wh_types.erl needs, around line 123:
```erlang
-ifdef(OTP_AT_LEAST_18).
-type sup_start_flags() :: supervisor:sup_flags().
-else.
-type sup_start_flags() :: {supervisor:strategy(), non_neg_integer(), non_neg_integer()}.
-endif.
```

**Comment by fenollp** [comment]
@jamesaimonetti De-update your types please.
The whole idea of the `OTP_AT_LEAST_18` macro is to not have to update types and thus maintain 2 versions of the same code

**Comment by jamesaimonetti** [comment]
@fenollp Since we're removing r15 from the target builds, only using 17+, there's no need to maintain that compat hrl. Plus I don't want to run make with all the ERLC options prepended, totally ruins the flow.

**Comment by fenollp** [comment]
@jamesaimonetti I agree. Make a ticket

**Comment by fenollp** [comment]
@lazedo You like this?

**Pull request merged**